### PR TITLE
delegate ItemStack

### PR DESCRIPTION
**Pull request by Machine-Maker**
This essentially removes API-only itemstacks without changing ItemStack to an interface, that will come later.

Just adds a field in api-ItemStack that should always be a CraftItemStack which the api-ItemStack will wrap around. CraftItemStack **has** to override any methods on api-ItemStack that use this `craftDelegate` to avoid recursion loops.

---
Download the paperclip jar for this pull request: [paper-10852.zip](https://nightly.link/PaperMC/Paper/actions/artifacts/1609690518.zip)

### CODE REVIEW

File: patches/api/0484-delegate-ItemStack.patch [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+ 
+     @Utility
+     protected ItemStack() {}
+@@ -86,8 +84,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
+     @Deprecated
+     public ItemStack(@NotNull final Material type, final int amount, final short damage, @Nullable final Byte data) {
+         Preconditions.checkArgument(type != null, "Material cannot be null");
+-        this.type = type;
+-        this.amount = amount;
++        this.craftDelegate = Bukkit.getUnsafe().createStack(type, amount); // Paper - delegate
```
Comment (masmc05) [inline patches/api/0484-delegate-ItemStack.patch]: Why not use <https://jd.papermc.io/paper/1.20.6/org/bukkit/inventory/ItemType.html#createItemStack(int)> and make CraftItemType always make CraftItemStack?

Comment (Machine-Maker) [reply patches/api/0484-delegate-ItemStack.patch]: yep, that is better. I still haven't fully accepted that ItemType and BlockType actually exist. its amazing.

**Comment by Machine-Maker** [comment]
It's possible that ItemStack#equals and ItemStack#hashCode should also directly delegate to a craftitemstack. I think that's probably most correct.

**Review by Lulu13022002 — COMMENTED** [review]
This doesn't handle empty itemstack (ItemStack#empty) which are still full api stack without server side equivalent and will throw a NPE for most methods.

**Comment by Machine-Maker** [comment]
> This doesn't handle empty itemstack (ItemStack#empty) which are still full api stack without server side equivalent and will throw a NPE for most methods.

I think I should just be able to change empty to return `ItemStack.create(Material.AIR)`. It'd have a stack size of 1 however, should I instead add a bridge method to create a truly empty stack? just by returning a copy of nms-ItemStack.EMPTY?

**Comment by Lulu13022002** [comment]
Another issue is that asNMSCopy doesn't accept null itemstack anymore so doing /clear for example throw a NPE too via the PrepareItemCraftEvent.

File: patches/api/0478-Proxy-ItemStack-to-CraftItemStack.patch [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
++    // Paper start - add static factory methods
++    /**
++     * Creates an itemstack with the specified item type and a count of 1.
++     *
++     * @param type the item type to use
++     * @return a new itemstack
++     * @throws IllegalArgumentException if the Material provided is not an item ({@link Material#isItem()})
++     */
++    @org.jetbrains.annotations.Contract(value = "_ -> new", pure = true)
++    public static @NotNull ItemStack create(final @NotNull Material type) {
```
Comment (kennytv) [inline patches/api/0478-Proxy-ItemStack-to-CraftItemStack.patch]: of, also below

**Review by lynxplay — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Review by kennytv — APPROVED** [review]

**Pull request merged**